python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,294 @@
+    """
+    if partition_value_as_str is None:
+        return None
+    raise TypeError(
+        f"Cannot convert partition string to python built-in, type {primitive_type} not supported: '{partition_value_as_str}'"
+    )
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return partition_value_as_str.lower() == "true"
+
+
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(primitive_type, partition_value_as_str: str):
+    """
+    Note: Before attempting to cast the value to string, a validation happens to ensure that there are no fractional digits. For
+    example, an invalid value such as "123.45" will be converted to a decimal with digits (1, 2, 3, 4, 5) with an exponent of -2 and the last
+    2 digits will be inspected for non-zero values. An example of a valid value is "123.00".
+    """
+    if partition_value_as_str is None:
+        return None
+    _, digits, exponent = Decimal(partition_value_as_str).as_tuple()
+    if exponent != 0 and any(digits[exponent:]):  # If there are digits to the right of the exponent, raise if any are not 0
+        raise ValueError(f"Cannot convert partition value, value cannot have fractional digits for {primitive_type} partition")
+    return int(float(partition_value_as_str))
+
+
+@from_partition_value_to_py.register(FloatType)
+@from_partition_value_to_py.register(DoubleType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return float(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(StringType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return partition_value_as_str
+
+
+@from_partition_value_to_py.register(UUIDType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return uuid.UUID(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(FixedType)
+@from_partition_value_to_py.register(BinaryType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return bytes(partition_value_as_str, "UTF-8")
+
+
+@from_partition_value_to_py.register(DecimalType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
+    return Decimal(partition_value_as_str)
+
+
+@singledispatch
+def to_bytes(primitive_type: PrimitiveType, value: Union[bool, bytes, Decimal, float, int, str, uuid.UUID]) -> bytes:
+    """A generic function which converts a built-in python value to bytes
+
+    This conversion follows the serialization scheme for storing single values as individual binary values defined in the Iceberg specification that
```

Comment:
Do we need to wrap lines in docstrings like this? It seems long?

Reply:
Good catch I would have expected the auto-formatter to chop this line. Fixed!